atom/browser/api/atom_api_system_preferences_mac.mm

```diff
@@ -438,6 +443,59 @@ AVMediaType ParseMediaType(const std::string& media_type) {
   }
 }
 
+bool SystemPreferences::IsTouchIDAvailable() {
+  if (@available(macOS 10.12.2, *)) {
+    base::scoped_nsobject<LAContext> context([[LAContext alloc] init]);
+    return [context
+        canEvaluatePolicy:LAPolicyDeviceOwnerAuthenticationWithBiometrics
+                    error:nil];
+  }
+  return false;
+}
+
+void OnTouchIDCompleted(scoped_refptr<util::Promise> promise, bool success) {
+  promise->Resolve(success);
+}
+
+v8::Local<v8::Promise> SystemPreferences::PromptTouchID(
+    v8::Isolate* isolate,
+    const std::string& reason) {
+  scoped_refptr<util::Promise> promise = new util::Promise(isolate);
+  if (@available(macOS 10.12.2, *)) {
+    base::scoped_nsobject<LAContext> context([[LAContext alloc] init]);
+    base::ScopedCFTypeRef<SecAccessControlRef> access_control =
+        base::ScopedCFTypeRef<SecAccessControlRef>(
+            SecAccessControlCreateWithFlags(
+                kCFAllocatorDefault,
+                kSecAttrAccessibleWhenUnlockedThisDeviceOnly,
+                kSecAccessControlPrivateKeyUsage |
+                    kSecAccessControlUserPresence,
+                nullptr));
+
+    scoped_refptr<base::SequencedTaskRunner> runner =
+        base::SequencedTaskRunnerHandle::Get();
+
+    [context
+        evaluateAccessControl:access_control
+                    operation:LAAccessControlOperationUseKeySign
+              localizedReason:[NSString stringWithUTF8String:reason.c_str()]
+                        reply:^(BOOL success, NSError* error) {
+                          if (!success) {
+                            promise->RejectWithErrorMessage(std::string(
+                                [error.localizedDescription UTF8String]));
```

Comment:
Does it matter what thread the promise is rejected in? This code block "is evaluated on a private queue internal to the framework in an unspecified threading context" [(source)](https://developer.apple.com/documentation/localauthentication/lacontext/1514174-evaluateaccesscontrol?language=objc)

Reply:
I think in this case it'd be fine? Chrome did [post it back in the original thread](https://cs.chromium.org/chromium/src/device/fido/mac/touch_id_context.mm?sq=package:chromium&g=0&l=112) but i don't think we can correspondingly do that with promises

Reply:
This is resolved by posting to `OnTouchIDFailed()`